nifi-commons/nifi-utils/src/main/java/org/apache/nifi/stream/io/MaxLengthInputStream.java

```diff
@@ -0,0 +1,128 @@
+ *  this work for additional information regarding copyright ownership.
+ *  The ASF licenses this file to You under the Apache License, Version 2.0
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.apache.nifi.stream.io;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+public class MaxLengthInputStream extends InputStream {
+
+    private static final String ERROR_MESSAGE_FORMAT = "Unable to read past maximum allowed length of %s bytes";
+
+    private final InputStream in;
+    private final long limit;
+    private long bytesRead = 0;
+    private long markOffset = -1L;
+
+    /**
+     * Constructs an input stream that will throw an exception if reading past a maximum length of bytes.
+     *
+     * @param in the underlying input stream
+     * @param limit maximum length of bytes to read from underlying input stream
+     */
+    public MaxLengthInputStream(final InputStream in, final long limit) {
+        this.in = in;
+        this.limit = (limit + 1);
+    }
+
+    @Override
+    public int read() throws IOException {
+        if (bytesRead >= limit) {
+            throw new IOException(ERROR_MESSAGE_FORMAT.formatted(limit));
+        }
+
+        final int val = in.read();
+        if (val > -1) {
+            bytesRead++;
+        }
+        return val;
+    }
+
+    @Override
+    public int read(final byte[] b) throws IOException {
+        if (bytesRead >= limit) {
+            throw new IOException(ERROR_MESSAGE_FORMAT.formatted(limit));
+        }
+
+        final int maxToRead = (int) Math.min(b.length, limit - bytesRead);
+
+        final int val = in.read(b, 0, maxToRead);
+        if (val > 0) {
+            bytesRead += val;
+        }
+        return val;
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        if (bytesRead >= limit) {
+            throw new IOException(ERROR_MESSAGE_FORMAT.formatted(limit));
+        }
+
+        final int maxToRead = (int) Math.min(len, limit - bytesRead);
+
+        final int val = in.read(b, off, maxToRead);
+        if (val > 0) {
+            bytesRead += val;
+        }
+        return val;
+    }
+
+    @Override
+    public long skip(final long n) throws IOException {
+        final long toSkip = Math.min(n, limit - bytesRead);
+        final long skipped = in.skip(toSkip);
+        bytesRead += skipped;
+        return skipped;
+    }
+
+    @Override
+    public int available() throws IOException {
+        return in.available();
```

Comment:
I'm leaning towards leaving this unchanged since we want the caller to keep reading and encounter the error if the available data exceeds the limit. There is a similar class in nifi-utils called `LimitingInputStream` which was a similar concept, except it just keeps returning `-1` once reaching the limit, and that one seems to also not alter the available logic.

Reply:
Got it, that makes sense.